Initial support proposal for http2

phouri opened the pull request:
Hi,

PiniH here - used wrong account before (work...).

Regarding this issue: https://github.com/expressjs/express/issues/3388

This is an initial proposal for making http2 work, this requires node to expose the Http2ServerRequest/Response and creates both request/response on the app and attaches the correct one depending on the request.

This is just an initial proposal (and a working example if Http2ServerRequest/Response) is exposed.

PiniH committed:
Initial support proposal for http2

phouri commented:
@dougwilson once I clear the PR on nodeJS I will continue working on this if you'd like.

My gut feeling is that there will be a need to split it into 2 separate files - httpRequest and http2Request (I just did a quick work here to make it work, I don't think it's the right approach) - and take out all the special decorators into another file and use them accordingly.

WDYT?

PiniH commented:
@dougwilson Hey, any thoughts regarding this? now that the PR in node is in we can move forward on this end :)

dougwilson commented:
Seems to be in the right direction. Probably need to have CI run test suite twice: once for HTTP/1 and once for HTTP/2.

Also the test suite here does leave a lot of the fine details up to the dependencies like send, on-headers, on-finished, finalhandler, etc. and they all need their test suites run against thr HTTP/2 server to make sure they still work right.

dougwilson commented:
P.S. I added the "5.x" label for now since this PR is changing exported API in a non-backwards-compatible way. That's not a bad thing, becasuse a 5.0 for this is fine, just wanted to note the reason.

dougwilson commented on lib/request.js:
So this is not going to make a compatible API. In Express, req.path is the pathname of thr URL (which, for instance, does not include the query stribg). Th HTTP2 prototype just aliases req.url to req.path, meaning they are the same and the req.path for HTTP2 would include the query string.

If Node.js code is going to insist on keep this change, then this could never land prior to 5.0 and req.path needs to go through a drprecation cycle in 4.x as well. Then this implementation can ignore req.path.

phouri replied:
I see, 

What if we override this here and return the pathname like http1 does? 
This will however create inconsistencies between http2 and http1, not sure what's better.

dougwilson replied:
If it's possible to do, absolutely! If you can't, 
please let James in thr Node.js issue know as well. Consistency in the Express.js API is what you are promised when you use Express.js, so I'm not really understanding what inconsistency you mean would come from this. Can you explain more?

phouri replied:
Actually the inconsistency I thought about would be between the IncomingMessage and Http2ServerMessage but you are right - this is not an issue if you use express and you expect express stuff :)

I will add an override that will return the pathname like normal req.path does in express in the weekend.

phouri replied:
I will also work on reverting the api change, get all the common decorators into a util helper and create new requestHttp2.js files instead of same file returning both requests/responses.

dougwilson replied:
Ah, gotcha. Yea, even the http1 class has slightly changed across Node.js versions. To that end we don't document the Node.js given APIs and just tell users to refer to the docs for their Node.js version.

dougwilson replied:
Regarding the refactor: cool. The changes made to the test/exports file is the indication that the PR was not backwards-compatible, if you're wondering how I could tell :)

phouri commented:
I tried looking up how to make the test utils use http2 client instead of http1, seems like it's not supported at the moment: https://github.com/visionmedia/supertest/issues/429

As to breaking the API - I thought that maybe I can rewrite this to make separate request and requestHttp2 files - and put the common decorators in a common code, that way it won't break existing API.

As to increasing the testing suites, I think I will need a little help with that :)

phouri commented:
Yep, reverted all the changes I did to the test, expect a pr tomorrow, still not sure how to accomplish all the tests.

phouri committed:
Seperate to files and revert breaking api

phouri committed:
Require parseurl and don't emit warning on setter

phouri committed:
Lint

phouri commented:
Finished the refactor for this.

Moved all the code that was inside request.js / response.js to the requestDecorator/responseDecorator.

request.js and response.js now again returns the httpRequest without breaking the api.

Adjusted the req.path on http2 to return only pathname - had to create a new mini object for the parseurl library (it only ever used the url part of the request, and cached the results).

@dougwilson  LMK what you think

phouri committed:
Http2 Example

phouri committed:
Lint...

benjamingr commented on lib/http2Request.js:
```diff
@@ -0,0 +1,45 @@
+ * as it returns full url
+ * Gets the full URL from the headers
+ * Creates a mock object for parse to cache on and read from
+ * as url in http2 refers to this.path which will results in an error
```
@jasnell this looks fine to me for compatibility, pinging you in to take a look :)

dougwilson replied:
As noted below, I don't think this is something we want to do; one of the key aspects od Express (and this is so key that all official middleware are written this way) is that you can write anything using onlt the base type and you can plop that in Express and it'll work 100% the same. If you look like pretty much all our middleware, we don't even test them against Express because of this core philosophy.

benjamingr commented on lib/http2Response.js:
```diff
@@ -0,0 +1,40 @@
+ * Remove if node adds deprecation setter as well
+ */
+Object.defineProperty(res, 'statusMessage', {
+  set: function (x) {
```
I had no idea you can define a setter separately from a getter, but apparently this works. 

This can be a `set() {}` by the way (shorthand property)

benjamingr replied:
Nevermind, it can't, I forgot this needs to run on old Node since it's not versioned with core. Sorry for the misleading comment!

dougwilson replied:
We should just wait for the setter to be in Node.js core, as the wrapper claims to be compatible for you only use the public API and setting the statusMessage is public API.

phouri replied:
I will try to see what to do about the status message on the node part then.

However one thing I noticed is that express always tries to set
statusMessage which if node add setter will result in deprecation message
if you use express with http2 - any ideas how to avoid that?

dougwilson replied:
I would say just change where it is being set to base that on the http version property of the request.

dougwilson replied:
The change to Node.js core add a setter looks like it will land very soon (PR https://github.com/nodejs/node/pull/15193) so we won't nee the custom setter any longer.

phouri replied:
Awesome, will remove it then.

stigok replied:
Landed in [Version 8.5.0](https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V8.md#2017-09-12-version-850-current-mylesborins)

benjamingr commented on lib/requestDecorator.js:
```diff
@@ -0,0 +1,502 @@
+ * express
+ * Copyright(c) 2009-2013 TJ Holowaychuk
+ * Copyright(c) 2013 Roman Shtylman
+ * Copyright(c) 2014-2015 Douglas Christopher Wilson
```
@dougwilson can you confirm this copyright header is fine?

dougwilson replied:
If it correctly represents the copyrights of the contents that were moved then it's fine. I can dig in sometime after the holiday if it is an open question.

benjamingr commented on lib/utils.js:
```diff
@@ -274,6 +288,23 @@ exports.setCharset = function setCharset(type, charset) {
+ * @private
+ */
+
+exports.defineGetter = function defineGetter(obj, name, getter) {
```
I'm not sure if we really need a utility for this so +0 on this change

dougwilson replied:
Yea, I agree, no utility is necessary.

phouri replied:
Ok, I'll take it out then.
Although I just copied it from the request response instead of repeating.

benjamingr commented:
Code looks really nice and design looks pretty well thought out. 

Would be really nice to see tests pass on http2 too :)

dougwilson commented on examples/http2/index.js:
```diff
@@ -0,0 +1,40 @@
+var isHttp2Supported = require('../../lib/utils').isHttp2Supported;
```
The examples should not be reaching into Express internals.

benjamingr replied:
I think this can (and should) be exposed as a public API (isHttp2Supported)

dougwilson commented on examples/http2/index.js:
```diff
@@ -0,0 +1,40 @@
+var isHttp2Supported = require('../../lib/utils').isHttp2Supported;
+if (isHttp2Supported) {
```
...and sincd this is a http2 example, just remove this wrapper.

dougwilson commented on examples/http2/index.js:
```diff
@@ -0,0 +1,40 @@
+    pushStyle(res);
+    next();
+  });
+  app.get('/index.html', function(req, res, next) {
```
I see you are trying to show an example of pushing thr dtyle sheet for html. You missed /index works as well. Thr static middleware has a callback you can use to see the file it is about to send, which would be a better example instead of trying to enumerate them as routes.

phouri replied:
I didn't know about the static callback :) 
How do I set it up? did you mean the setHeaders callback?

dougwilson replied:
Sorry, on holiday and wasn't easy to lookup :) so the setHeader could be used, but seems to really be an abuse of it. Looking into it, what I was thinking was the "file" event that is in send. So I wonder if as part of getting serve-static/send/deps working with http2, another callback should be added in order to make pushing file assets very straight forward.

phouri replied:
Adding support for the new http2 push is something I was wondering how to do, because I think the best way express could solve it is to make it go through another route when you hit one route you know would require the other. i.e. when requiring index.html send it also to style.css (assuming no static is used), but I have no clue where to begin ;) 

How do  I use that file event from send? I thought it was pretty deep inside.

dougwilson commented on lib/http2Request.js:
```diff
@@ -0,0 +1,45 @@
+    return '';
+  }
+  path = headers[http2Constants.HTTP2_HEADER_PATH];
+  this.parsedObject = this.parsedObject || {url: path};
```
The req.path should reflect any changes made to req.url over thr course of the request (for example middleware cut thr matched part). This is accidentally caching the first access value.

phouri replied:
It does not, if the url is set, it will rewrite the path, which in turn will rewrite the url on mock object and cause parse url to rerun.

dougwilson replied:
You are correct, I see the object alteration below, sorry! It would be awesome if it could just use the same cache that the routing used, but if this is the only way, then we'll just have to live with the perf hit req.path will get.

phouri replied:
The only performance hit (to the best of my understanding is the creation of the new object - otherwise it uses the same caching mechanism and everything works exactly the same way, instead of operation on the req object though it operates on a new (only created once) object, and the cache is set on it.

dougwilson replied:
That's not the case here. The router itself in Express is prepopulating the cache, but this req.path is using a separate cache, requiring a double parse. Since the movement through middleware will keep altering req.url, the internal prepopulation the router is doing for that is not getting picked up here. For example, the router will populate the cache and then in a middleware you call req.path it has to parse again instead of reusing the cache. Then the request moves on, the router alters req.url and saves the new parse in the cache. The call to req.path again requires yet another reparse since it uses a separate cache, etc. etc.

dougwilson reviewed (changes requested):
Couple notes above, and I have no idea how we are going to get this merged into the 5.0 branch, since this rewrites files heavily changed in the 5.0 branch. Any thoughts on here would be good. Maybe at a certain point prior to merging into r, you can stage a merge into 5, idk

And of course we need to get the test suite running under http2 somehow. We use thr aupertest module currently, which uses superagent. I assume that either (1) those will need to get http2 supported or (2) rewrite our tests with a framework that does support it.

phouri commented:
Btw, about the 5 branch - I removed all breaking changes I did in previous commits, perhaps this could even go into the 4 branch as is, and I will try to see what changes are needed in a different PR for 5 branch.

benjamingr commented:
It would be amazing for the Node.js project if this landed somehow in 4.x, even if the documentation tags it as "experimental".

dougwilson commented:
Unless Node.js is willing to remove the req.path field, this is just never going to land in 4.x. Express id built around not altering any base behavior, only adding new helpers. Since req.path in 4 works differently than req.path in the http2 compat layer, it means that you cannot just take code that works on the http2 compat layer and place it inside Express, because the value of req.path may differ now.

benjamingr commented:
@dougwilson I agree, but I think feedback from users will be essential for a good HTTP2 story. I think we can expose HTTP2 turned off by default with a flag to turn it on in express and ship like that - or not make any claims about express supporting it at all but allow passing express to an HTTP2 server listener without saying it is officially supported.

dougwilson commented:
... Express has no conxept of feature switches. Previews and feedback are done using alphas, betas, RCs, etc.

dougwilson commented:
For non published things, we ask for install and feedback by providing instructions to install from a branch in the repo. We can merge this into a http2 branch and tell folks to npm install expressjs/express#http2 as well.

dougwilson commented:
The other issue with 4 is that the app.request and app.response (and global express.request and express.response) won't work for http2 as implemented here. This is the way in which folks are supposed to attach their own sugar for their app or globally, but anything they add there won't be there for http2, which still means that folks will need to change their apps to work with http2. I know not a big deal, but certainly just another thing that would make it harder for people to forklift their existing apps to http2. I bring this up because I am guessing that is what your thought on why 4 is so important to land in? To me, 5.0 is the most appropriate. If Node.js landed http2 in Node.js 6 (current LTS) then that would be the equivalent of landing this in Express 4. Node.js can just land http2 in Node.js 6 turned off by default with a flag to turn it on and ship like that.

phouri committed:
Fix example

dougwilson commented on examples/http2/index.js:
```diff
@@ -0,0 +1,38 @@
+try {
```
Remove this try wrapper. This is a http2 example, ideally the examples are how folks will write their code. Do you think everyone will be wrapping their servers in try blocks when they start using http2?

phouri replied:
Aren't the tests running the examples? Maybe I was under a false impression here

phouri replied:
Nm, my bad, removing :)

phouri committed:
Remove try/catch in example
Better override for path/url - use same mechanism in http1, make sure to bring full path for url

phouri commented:
Regarding the 4/5 debate, I am the new guy here so take it with a grain of salt.

Imo - if I was a company already having express 4 running as my webserver and hearing that support for http2 is available (this doesn't break anything for existing users) - I'd be more than happy to see if switching to http2 just worked, and if there would be some remark - that if you decorated app.request/response you have to decorate app.http2Request/Response I think it might be worth it for many users, as upgrading to 5 sounds like it'd bring a lot more pain. 

Just my opinion, if we decide to go with 5 I will make sure to do the required changes.

phouri commented on lib/http2Request.js:
```diff
@@ -0,0 +1,40 @@
+ * Creates a mock object for parse to cache on and read from
+ * as url in http2 refers to this.path which will results in an error
+ */
+defineGetter(http2Req, 'url', function path() {
```
@dougwilson 
I found a better solution, in retrospect my solution was bad :)

I changed it to basically work the same, except that I rewrite the getter for url to give the raw path (avoiding the endless loop on get path / url until it explodes).

Basically what happens in http2:
get path -> overridden by the decorator to return parse object pathname
set path -> sets the full path in the headers
get url -> brings the full path from headers (instead of this.path as in the unedited Http2ServerRequest)
set url -> sets the path

dougwilson replied:
Looks better, but i think the function should be named url instead of path. We also are not going to land sugar functions that monkey patch the underlying http module regardless. Express 3 and prior used to do that and it was a disaster in the long run, because it makes assumptions based on the inner implementation of Node.js core. The HTTP2 compat layer really makes it scary to even attempt to make the implementation assumptions this makes. The docs from Node.js:

> This API targets only the public API of the HTTP/1, however many modules uses internal methods or state, and those are not supported as it is a completely different implementation.

Does Node.js think that replacing the setters is something that is sustainable? Can this be noted in the docs?

In fact https://nodejs.org/dist/latest-v8.x/docs/api/http2.html#http2_class_http2_http2serverrequest doesn't even document req.path. Why are we working around something undocumented in Node.js? Seems very dangerious.

phouri replied:
Oh, right, forgot the function.

You are right - this does seem a bit dangerous, perhaps some opinion from Node.js team as to what we can do here would be best. Unfortunately without this change it just doesn't work.

@benjamingr @jasnell

benjamingr replied:
@nodejs/http2

dougwilson commented:
Yea, but it just isn't likely to happen. We knew a while back when planning 5.0 that http2 would likely not be compatible to the choices made in Express 4. The spdy module actually worked to implement http2 in a compatible way with Express. It is sad that Node.js core did not, and that pretty much leaves this as either (1) Node.js core should make the http2 compatible with Express just as spdy did or (2) this is an Express 5 change.

The biggest issue is that Express 4 is fundamentally tied to the single parent in the prototype, which sucks really bad. This needs to be fixed, but it will only bring pain and confusion to users if this is jammed into Express 4.

In addition, Node.js has even introduced req.path, a conflict with our existing sugar that doesn't work the same. This is just not going to allow us to lans support in Express 4, and the reasons for this are noted above.

phouri committed:
Remove defineGetter from utils

phouri commented:
Regarding the testing:

I played around with settings supertest to work with http2, it actually ran but failed in numerous places (getting headers etc), I will continue my investigation to see perhaps it's possible to add support for http2 and I will create a PR for supertest.

phouri committed:
Rename function name

benjamingr commented:
>  It is sad that Node.js core did not, and that pretty much leaves this as either (1) Node.js core should make the http2 compatible with Express just as spdy did or (2) this is an Express 5 change.

@jasnell and `@nodejs/http2` certainly made efforts and kept express in mind - maybe we can schedule a hangout next week about what changes need to happen?

phouri committed:
Remove overridding setter for statusMessage - will land in Node soon

dougwilson commented:
Hi @benjamingr I'm happy to have a hangout and just as happy to land this in Express 5. Like I said, Express 4 is fundamentally tied to only working with objects that derive from IncomingMessage. The `spdy` module was happy to make their object derive from that, and that's pretty much the only possible road to land this in Express 4. If that isn't going to happen, I'm happy to land this in Express 5 just the same. I don't understand what the pushback is to Express 5; HTTP/2 was the biggest reason we have been holding Express 5 for, because we figured that Node.js core was not going to go down the route that `spdy` did -- and it didn't.

benjamingr commented:
Thanks @dougwilson,

I'll explain my stance: I think that express is a huge consumer of HTTP/2.0 in Node.js and I'd like to make sure that the considerations of express are given a high priority while the API is still experimental behind a flag.

Once the API ships without a flag then the API is mostly set in stone. I think that all sides are willing to work towards a good express http2 story - and we have @phouri who is willing to make adjustments on either side.

I figured a hangout would be a good way to coordinate the efforts since I am confused at the moment about where things stand. I acknowledge that it's entirely possible that everyone else knows what's going on and it's simply my failure to keep track of things - so up to you.

dougwilson commented:
Hi @benjamingr makes sense. As it stands: Node.js core needs to rewrite the compatibility API to inherit the existing `http.IncomingMessage` class instead of an independent class in order to work with Express 4 or we'll make adjustments to land it in Express 5. We have been planning the Express 5 route for a long time now, and imo that is fine.

I'm just confused why there seems to be push back on landing this in Express 5, given that the current way in which the Node.js core http2 is designed (and the way the npm `http2` module was designed, btw) is incompatible with Express 4 design. I would have been happy to point this out and help with the http2, but I didn't even realize the Node.js core http2 process was even happening until it was just about to land in Node.js 8, so I don't want to force Node.js to change everything after the fact and would rather just give in to the Node.js implementation and get the necessary changes in Express 5 to support them.

dougwilson commented:
And of course the `req.path` issue: Node.js code added `req.path` to the HTTP/2 compatibility API, with an implementation that is incompatible with the Express.js `req.path` implementation. Node.js core would need to remove it, or Express 5 will just remove it as proposed in #3407 , but that is the other Express 4 adoption blocker (though not as big as the prototype chain issue).

benjamingr commented:
> I'm just confused why there seems to be push back on landing this in Express 5

The push back isn't really about landing this in Express 5 - it's more about the API breakage itself. I'm concerned that it might affect adoption and migration of Express 4 projects.

I'll wait for James to respond and we'll coordinate :) 
I'll open a corresponding issue at Node.js/node.

mcollina commented:
I designed the compatibility layer. So, part of this is my fault. I'm happy to have a call whenever you like. Fire up a doodle and I'll be joining.

> As it stands: Node.js core needs to rewrite the compatibility API to inherit the existing http.IncomingMessage class instead of an independent class in order to work with Express 4 or we'll make adjustments to land it in Express 5. We have been planning the Express 5 route for a long time now, and imo that is fine.

I still do not fully understand how Express handles inheritance from core. Why do you need to inherit from `IncomingMessage`? How do you feed the new objects into the rest of the framework? The basics problem here is that all the properties of the compat layer are extracted from some other internal data structures in HTTP2. We need those data structures for HTTP2 to be fast and reliable, and we iterated on that design quite a bit before landing on the current structure - we plan to change it further to reduce buffering and improve throughput.

On the good side, we can change all the methods of the http2 compat layer to whatever it's needed. That is no big deal.

Another option is to semver-major break `IncomingMessage` itself but I do not think that is a good idea. (I am happy to be proven wrong, so feel free to send a PR). 

> And of course the req.path issue: Node.js code added req.path to the HTTP/2 compatibility API, with an implementation that is incompatible with the Express.js req.path implementation. Node.js core would need to remove it, or Express 5 will just remove it as proposed in #3407 , but that is the other Express 4 adoption blocker (though not as big as the prototype chain issue).

Let's remove it. All of this is experimental anyway.
I think it was added because everybody is so familiar with Express that we thought it was part of the Node.js API.

The change should be trivial: https://github.com/nodejs/node/blob/ad3d2ce68a92fd160f72ed5c0dba0e18a90e906d/lib/internal/http2/compat.js#L220-L240

dougwilson commented:
Hi @mcollina what is a doodle?

mcollina commented:
@dougwilson http://doodle.com to coordinate meeting times across multiple people and different timezones. Sorry I use that all the time.

phouri commented:
If it's ok with you guys, I'd be happy to join and help the effort :)

jasnell commented:
Yeah, compatibility is something that I went back and forth on for months while putting the design of this together. Unfortunately, the more I stressed full compatibility, the worse the performance got, so we ended up with a meet in the middle hybrid approach. I think we can get there but it'll take some work for certain. Happy to help! :-)

dougwilson commented:
I assume anyone can join. I'll make this a Hangout that will be recorded to the Express.js YouTube channel, unless there is a reason not to do so. I created https://doodle.com/poll/nrxufh42ytg8xhkg but I don't know what dates and times to choose.

mcollina commented:
@dougwilson anything weekdays during business hours or evenings (CEST timezone) work for me. Weekends is a bit more problematic right now.

dougwilson commented:
Hi @mcollina haha, I am exactly the opposite right now :) I can set this a few weeks out if that's OK.

dougwilson commented:
For those on the Doodle thing, I just found out today that that week I put the dates for I have to travel to San Francisco. This means that I may not actually be available then, and need to determine exactly what the trip entails. @mcollina noted the best times for him (anything weekdays during business hours or evenings (CEST timezone)). Maybe if there is just some general things I can make a new Doodle based on those.

mcollina commented:
I can be free on the 23rd.

Why don't we start opening an issue for all the blockers in the meanwhile?

phouri committed:
Remove unneeded decorators after node js fixes

apapirovski commented:
Correct me if I'm wrong but it seems like at this point the main need here would be around porting all the tests in express to support http2? @phouri have you done any work in that regard? It seems like adding `node-spdy` support in `supertest` would be the best route since it has an equivalent API to `http`.

I'm looking at doing some work on this so wanted to make sure I'm not duplicating any work already being done.

(There has been a lot of work on the compatibility API in http2 so it should be getting a lot closer to being usable interchangeably with http.)

phouri commented:
I tried looking at adding support to supertest for http2, but didn't dig too deep, spdy sounds like an awesome idea but.. if we are already doing work with the native http2 module perhaps would be better but more complicated.




Also I think another issue here is that merging this into 4 wouldn't work completely out of the box as extending the app prototype would have to be done with consideration for http2. 




If you want/need help with the supertest part lmk , I'd love to finish this PR.




On Tue, Sep 19, 2017 at 5:56 PM +0300, "Anatoli Papirovski" <notifications@github.com> wrote:












Correct me if I'm wrong but it seems like at this point the main need here would be around porting all the tests in express to support http2? @phouri have you done any work in that regard? It seems like adding node-spdy support in supertest would be the best route since it has an equivalent API to http.


I'm looking at doing some work on this so wanted to make sure I'm not duplicating any work already being done.



—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

apapirovski commented:
I have to imagine getting this into Express 5 is probably fine since I don't think there's any way http2 will inherit from `http.IncomingMessage` & `http.ServerResponse`.

http2 client doesn't have support for the whole Agent model while `node-spdy` does. I think down the line using the native implementation would be nice but I think whatever gets us there in the meantime is probably good enough.

mcollina commented:
IMHO we might need a bit more support for core to add supertest http2 capabilities, as the core client is very crude atm.

phouri commented:
Any issues open on the core client that I can help with? Or some pointers?

On 19 Sep 2017 18:12, "Matteo Collina" <notifications@github.com> wrote:

> IMHO we might need a bit more support for core to add supertest http2
> capabilities, as the core client is very crude atm.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/expressjs/express/pull/3390#issuecomment-330571478>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE7wospVKhShrnXjVR3Id0o2hC66P4Awks5sj9nKgaJpZM4OwldU>
> .
>

mcollina commented:
There is a discussion going on in https://github.com/nodejs/node/issues/15271.
The current API that we have is the equivalent of the "core" api on the server. That might be enough, or might not.

apapirovski commented:
I hacked together a local version of superagent & supertest that runs all the tests with http2 and I get 64 failing test cases (771 passing). Honestly, that's a lot better than I had anticipated. I'm going to work on making it reusable and also dig into what the failures actually are, and whether it's issues with node or just fundamental differences between h1 & h2.

dougwilson commented:
@apapirovski that's awesome! If you have those changes anywhere I can access, I can always take a look at the failing tests as well.

apapirovski commented:
Will post more details this coming week, when I have a moment. 64 was wrong, it was actually 75 which I've been able to get down to 62 by working on the http2 compatibility layer. 

There are like 30 errors around host & subdomain stuff which I'll need to dig into. The rest are related to `app.response` not being the same thing as `app.http2Response`. I think any final solution will need to (unfortunately) be more complex than the one currently presented in this PR so that `app.response` & `app.request` can be modified and reflect on both h1 & h2 prototypes.

apapirovski commented on lib/middleware/init.js:
```diff
@@ -31,9 +35,14 @@ exports.init = function(app){
     req.res = res;
     res.req = req;
     req.next = next;
+    if (isHttp2Supported && req instanceof http2Request) {
```
This should be checked by looking at `req.httpVersionMajor ` instead.

phouri replied:
Not sure about this: 
What if https://github.com/spdy-http2/node-spdy/issues/316 is indeed fixed yet node-spdy still uses the http1 IncomingMessage?

apapirovski replied:
This actually seems fine now that I have fresh eyes. I don't know why I thought it was an issue anyway, I guess just stared at the code too long.

phouri commented:
@apapirovski  That's awesome! 

Any chance you could indeed share the tweaks so I can run it as well?

I tried doing a more sophisticated solution - making the request/response prototype agnostic and just set the methods, you can take a look here: https://github.com/phouri/express/tree/request-response-agnostic-http2 the reason I didn't mention it / go through with it is the benchmark too quite a hit, from ~8000/s to ~5700/s which to me doesn't sound like something we can live with. 

If you have any suggestions on how to do this quickly that'd be awesome. 

The prototype chain should be something like:  decorator -> request prototype (which would be dynamically set at runtime), I couldn't figure out a way to do it quickly.

apapirovski commented:
Here's a [dump of all the errors](https://gist.github.com/apapirovski/fe8768a4dab535155270d869a9e24a8a) after I've fixed all the ones that seemed to be straight-up `http2` bugs. A bunch seem like they might be `spdy` "bugs" (like the headers & cookies ones) and others are because not all functionality works properly (since `app.request` & `app.response` refers to http1). All in all, not too bad.

It'll be a a few days, maybe a week before I can get something together so other people can run the tests. I just hacked my local npm installs. It's not too hard but it's little changes here and there, all over the place.

Edit: I've [updated](https://gist.github.com/apapirovski/fe8768a4dab535155270d869a9e24a8a) the gist. We're down to 32, 6 of which are a spdy issue (9, 10, 11, 12, 13, 14) and the vast majority of the rest also don't seem like strictly `http2` issues. I'm still unclear on 1.

My next step is going to be working on a patch for superagent/supertest or just something that can be used to run the tests. Will see if I can get the client in core to do enough, if not I might need to fix bugs in spdy (and the libraries it depends on) first. Will post again when I've got that done.

phouri commented:
@apapirovski nice! For all the proto inheritage etc, could be nice to run the tests by flipping the request/response (i.e. app.request will be http2Request and app.http2Request will be normal IncomingMessage) just to nullify these tests failing and attribute them to the change in the prototype. Could also change the tests to mutate both http1 and http2 protos.

1 is indeed curious - could it be attributed to how errors are propagated in http2 module?
2-7 sound like are just the proto enhancing and were expected.
22-30 sound a lot like the prototypes got mixed up (that was the error I first encountered when just wiring express and http2 module for the first time)

Not sure about all the rest.

Waiting for your updates! :)

dougwilson commented:
For number 1, the test relies on the error `Error: Can't set headers after they are sent.` occurring on `res.setHeader` after a `res.writeHead` call (yes, I know, but a lot of these tests are driven off real issues encountered over the years by Express users). In other words, the following is expected to error:
```
$ node -e 'h=require("http"),s=h.createServer((req,res)=>{res.writeHead(200);res.setHeader("foo","bar")}).listen(0,()=>h.get(`http://localhost:${s.address().port}`))'
_http_outgoing.js:357
    throw new Error('Can\'t set headers after they are sent.');
    ^

Error: Can't set headers after they are sent.
    at ServerResponse.OutgoingMessage.setHeader (_http_outgoing.js:357:11)
    at Server.h.createServer ([eval]:1:71)
    at emitTwo (events.js:106:13)
    at Server.emit (events.js:191:7)
    at HTTPParser.parserOnIncoming [as onIncoming] (_http_server.js:546:12)
    at HTTPParser.parserOnHeadersComplete (_http_common.js:99:23)
```

dougwilson commented:
The one for number 5 indicates that during the request the `req.url` became `undefined` somehow. I know this is a getter for the `:path` header, so possibly the header became unset or something?

apapirovski commented:
@dougwilson thanks for the pointer for 1. You can ignore 5 — that was just a prototype issue. I monkey patched everything that was left referencing http1 to use http2 and updated the gist. Down to 14 errors but 3-8 are definitely issues with `spdy`.

apapirovski commented:
I think I'm on my way to resolving almost everything but this one issue is probably not something we can fix for `http2`:

```
1) throw after .end() should fail gracefully:
     Uncaught TypeError: Cannot read property 'destroy' of undefined
      at Immediate.<anonymous> (node_modules/finalhandler/index.js:130:18)
      at Immediate._onImmediate (lib/router/index.js:635:15)
```

This is happening because `finalhandler` tries to call `req.socket.destroy()` if an error occurred but a) that's not possible with http2 compat (we have to give up our reference to the Http2Stream, etc.), and b) it's not even the correct behaviour for http2. Any thoughts @dougwilson?

dougwilson commented:
That library can likely fork what it does for http/2. What would be the correcr behavior for http/2?

Another awesome thing, if you don't mind, would be checking the on-finished rmtest suite against http/2. I can do that as well if you have somewhere I can pull down your work.

jasnell commented:
For http2, the correct thing would be to destroy the `Http2Stream` rather than the `Socket`. The `Socket` instance is completely taken over and owned by the `Http2Session` and may have multiple requests/responses in flight despite any error that may occur within any single request/response stream. Basically, with http2, userland code should *never* touch an http2 socket directly or else many bad things could happen.

apapirovski commented:
Just to expand on it, you would call `res.destroy(err)` in that case and it would take it from there (if the stream is still available).

Will try to get on-finished test suite running later and post an update here. I made some patches to `http2` compatibility to support it but it might not be 100% accurate yet.

apapirovski commented:
This is the final issue with `http2` in the cookies npm package:

```js
var setHeader = res.set ? http.OutgoingMessage.prototype.setHeader : res.setHeader
setHeader.call(res, 'Set-Cookie', headers)
```

Any thoughts @dougwilson? I'm guessing it could be updated together with v5?

apapirovski commented:
Hmm... `on-finished` will be a problem. I think it'll need to be updated to be `http2` aware. It relies too much on capturing events from the socket and also checking its status. As @jasnell mentioned, trying to glean anything from the socket with `http2` is pointless.

To be more specific, it works fine with `Http2ServerResponse` because `finished` can do most of the job but with the readable (request) side it's much harder. I don't think we'll be able to tweak the compatibility implementation to make that one work.

On the plus side, the compatibility implementation itself seems mostly equivalent to http1 now. Will open a PR later today in node so that anyone else that wants to can pull it down.

dougwilson commented:
Gotcha. The mechanisms of on-finished are critucal to the correct functuonality of real Express apps, which is why I pointed it out. Thr on-headers is another one as well. Because I split up the inner workers, just running the test suite of Express only tests the superficial functionality of the framework, as I hid away the critical functionality into the dependencies and their test suites so they could be shared with Koa as the main consumer, and anyone else as well.

dougwilson commented:
But if you can share how you are doing this testing, I can work on figuring some of this out as well, if you want help.

dougwilson commented:
Oh, and that linr from the cookies module was the necessary workaround for the monkey patxhibg that express 3 was doing (and I think old Restify). We can investigate either getting it to work for http2 as well, or probably make a new major version that doesn't do that crap any longer (better option).

mcollina commented:
Considering that HTTP2 is a very different protocol, I think on-finished should have a completely different code path for the two version of http.

If we can add things to the HTTP1 implementation to make on-finished simpler, we should seek that as well. Detecting when a HTTP request is done is too tricky.

dougwilson commented:
Right, it is tricky, which is why we took all the edge cases Express users have experienced iver the years and created a module that would actually solve them all. Adding more stuff to the HTTP/1 implementation is all well and good, but you can't change old versions, meaning that if maybe that was done back in Node.js 0.10 or even Node.js 4 we could use them, but we won't be able to exclusively rely on anything brand new in HTTP/1 for a long time to come.

I definitely don't disagree that http/2 should have a difderent code path, of course. But for the http1 compat layer, at what point is the line drawn on compatibility as well? Is the only issue adding listeners on the socket property or is it a bigger issue? We only added listeners there becuse the http1 implementation gives us no other choice.

mcollina commented:
You might attach things to the Http2Stream of the core API. It should work the same as your current socket.

We added .socket to the compat layer, but it is not useful at all. We were unsure when we added that, but it might be easier to detect if a socket is not there in the first place.
The socket is one of the areas were the new protocol would leak through. It is impossible to provide the same functionality there (as for other details, unfortunately the protocol is different).

dougwilson commented:
Right. Sorry, this is all really just speculation without understanding what the specific issues with `on-finished` are. Ideally if the tests could be shared, that would be amazing to get more eye balls on it to understand what is happening.

dougwilson commented:
FWIW, the `on-finished` has an open tracking issue for working with http/2 https://github.com/jshttp/on-finished/issues/27

apapirovski commented:
Sorry, I'll try to get something shareable soon. I literally went through and modified all the relevant npm packages in node_modules so it's kinda hard to share as is.

I'll also make a PR with an updated http2 test suite for on-finished so that anyone can run it and work on making it work with `http2.` That one is easier since it doesn't depend on anything.

> We added .socket to the compat layer, but it is not useful at all. We were unsure when we added that, but it might be easier to detect if a socket is not there in the first place.

I think it's best if it stays — at least for now — because, for example, the current http2 compat version works with hapi and without it, it would break. I don't know... it's all kinda tricky.

mcollina commented:
Are there any updates on this?

apapirovski commented:
Personally, haven't had time to create PRs for superagent & supertest which would be needed to get the test suite running without monkey patching node_modules. If anyone feels like doing that, please go ahead... 😄

phouri commented:
If you can send / post details on what you did I'd be happy to :)

dougwilson commented:
@apapirovski would you at least be able to provide the rough code you used? That would be a big help to me or anyone who wants to start a PR to those projects. You had to use something to determine all the things you did. Just a repo somewhere with the code would be really helpful.

dougwilson commented:
> do you want me to just make a git repo with my copy of express & node_modules? 😆I mean, I can do that... it'll be a mess to work with but I suppose it might be better than nothing.
> 
> I've been trying to carve out some time to get PRs for the actual packages (supertest, superagent, etc.)( used but it's a bit more work than just temporarily monkey-patching them.

@apapirovski that would work. I can always dig around in there and help as well. I definately get the time issue, haha. So just an unorganized dump I can volunteer time to make sense of it :+1:

wesleytodd commented:
Hey, I have not been involved in this discussion, just reading along mostly, but I was wondering if a better overall approach would be removing the prototype extension and replace it with our own custom `req`/`res` objects?

There has been discussion about doing that anyway, and it would greatly simplify this pull request I think.  What do people think about doing that before we finalize this?

phouri commented:
This PR was supposed to be a relatively small change that will also support http2. 
Refactoring req/res sounds like a big change, and will make this redundant 👍

wesleytodd commented:
Not redundant, just that the work you did here would also need to be applied there.  I am unsure if this could have been done in a way which didn't break backward compatibility, so it was probably going to only land in 5.x anyway.  With that in mind, it seems like resolving these then applying this on top is the best way forward:

https://github.com/expressjs/express/issues/2432
https://github.com/expressjs/express/pull/3214

AaronDiffenderfer-USDA commented:
@phouri any updates on when this PR will be finished and the http2 functionality available?

therealteekm commented:
bump @atdiff's question. 

if work on this is paused, what are some workarounds or alternative solutions that people are using?

dougwilson commented:
I'm happy to pick up the work that was left off. I asked if I could get a copy in https://github.com/expressjs/express/pull/3390#issuecomment-352191056 but never received anything so far. Anyone is welcome to pick this up as well, and I don't have a lot of time to put into this which is why I was really hoping to get that code.

sogaani commented:
Hello @dougwilson 
I'm working on this PR. 
[My forked branch](https://github.com/sogaani/express/tree/initial-support-http2) have been working tests with http2. Would you review my codes and help me to apply my commit on this PR?

Run tests with following commands.
```
npm install
npm run test-http2
```

I changed following packages to support http2.
We need to apply PR those packages first. 

* Other modified packages.
    * [superagent](https://github.com/sogaani/superagent/tree/http2)
    * [supertest](https://github.com/sogaani/supertest/tree/http2)
    * [cookies](https://github.com/sogaani/cookies/tree/http2)
    * [cookie-session](https://github.com/sogaani/cookie-session/tree/http2)
    * [vhost](https://github.com/sogaani/vhost/tree/http2)

benjamingr commented:
cc @phouri

phouri commented:
Feel free to take over this PR and continue what ever is needed, I unfortunately do not have time for this :(

phouri commented:
Can close this if needed

sogaani commented:
I create PR https://github.com/expressjs/express/pull/3730 to rebase and add tests.

Abourass commented:
So, now that Node 10 is LTS, and the HTTP2 module is included by default, what's going on with this?

dougwilson commented:
Since the pull request has merge conflicts and the author asked for it to be closed and a new pull request was created to take it's place, I am closing this now that it is effectively a duplicate and the author won't be moving it forward anyhow.
